File: go/indicators/core/indicator.go
package core

import (
	"zpano/entities"
)

// Indicator describes a common indicator functionality.
type Indicator interface {
	// IsPrimed indicates whether an indicator is primed.
	IsPrimed() bool

	// Metadata describes an output data of an indicator.
	Metadata() Metadata

	// UpdateScalar updates an indicator given the next scalar sample.
	UpdateScalar(sample *entities.Scalar) Output

	// UpdateBar updates an indicator given the next bar sample.
	UpdateBar(sample *entities.Bar) Output

	// UpdateQuote updates an indicator given the next quote sample.
	UpdateQuote(sample *entities.Quote) Output

	// UpdateQuote updates an indicator given the next trade sample.
	UpdateTrade(sample *entities.Trade) Output
}

// UpdateScalars updates the indicator given a slice of the next scalar samples.
func UpdateScalars(ind Indicator, samples []*entities.Scalar) []Output {
	length := len(samples)
	output := make([]Output, length)

	for i, d := range samples {
		output[i] = ind.UpdateScalar(d)
	}

	return output
}

// UpdateBars updates the indicator given a slice of the next bar samples.
func UpdateBars(ind Indicator, samples []*entities.Bar) []Output {
	length := len(samples)
	output := make([]Output, length)

	for i, d := range samples {
		output[i] = ind.UpdateBar(d)
	}

	return output
}

// UpdateQuotes updates the indicator given a slice of the next quote samples.
func UpdateQuotes(ind Indicator, samples []*entities.Quote) []Output {
	length := len(samples)
	output := make([]Output, length)

	for i, d := range samples {
		output[i] = ind.UpdateQuote(d)
	}

	return output
}

// UpdateTrades updates the indicator given a slice of the next trade samples.
func UpdateTrades(ind Indicator, samples []*entities.Trade) []Output {
	length := len(samples)
	output := make([]Output, length)

	for i, d := range samples {
		output[i] = ind.UpdateTrade(d)
	}

	return output
}

File: rs/src/indicators/core/indicator.rs
use crate::entities::bar::Bar;
use crate::entities::quote::Quote;
use crate::entities::scalar::Scalar;
use crate::entities::trade::Trade;

/// Output is a vector of boxed output values (Scalar, Band, Heatmap, Polyline).
pub type Output = Vec<Box<dyn std::any::Any>>;

/// Common indicator functionality.
pub trait Indicator {
    /// Indicates whether the indicator is primed.
    fn is_primed(&self) -> bool;

    /// Returns metadata describing this indicator and its outputs.
    fn metadata(&self) -> super::metadata::Metadata;

    /// Updates the indicator given the next scalar sample.
    fn update_scalar(&mut self, sample: &Scalar) -> Output;

    /// Updates the indicator given the next bar sample.
    fn update_bar(&mut self, sample: &Bar) -> Output;

    /// Updates the indicator given the next quote sample.
    fn update_quote(&mut self, sample: &Quote) -> Output;

    /// Updates the indicator given the next trade sample.
    fn update_trade(&mut self, sample: &Trade) -> Output;
}

/// Updates the indicator given a slice of scalar samples.
pub fn update_scalars(ind: &mut dyn Indicator, samples: &[Scalar]) -> Vec<Output> {
    samples.iter().map(|s| ind.update_scalar(s)).collect()
}

/// Updates the indicator given a slice of bar samples.
pub fn update_bars(ind: &mut dyn Indicator, samples: &[Bar]) -> Vec<Output> {
    samples.iter().map(|s| ind.update_bar(s)).collect()
}

/// Updates the indicator given a slice of quote samples.
pub fn update_quotes(ind: &mut dyn Indicator, samples: &[Quote]) -> Vec<Output> {
    samples.iter().map(|s| ind.update_quote(s)).collect()
}

/// Updates the indicator given a slice of trade samples.
pub fn update_trades(ind: &mut dyn Indicator, samples: &[Trade]) -> Vec<Output> {
    samples.iter().map(|s| ind.update_trade(s)).collect()
}
